plugins/synaptics-rmi/fu-synaptics-rmi-v5-device.c

```diff
@@ -118,19 +140,164 @@ fu_synaptics_rmi_v5_device_write_block (FuSynapticsRmiDevice *self,
+			g_autoptr(GByteArray) res = NULL;
+			res = fu_synaptics_rmi_device_read_packet_register (self,
+									    f34->query_base + 14, /* addr of flash properties + 5 */
+									    0x3,
+									    error);
+			if (res == NULL)
+				return FALSE;
+			if (res->len != 0x3)
+				g_debug ("read %u bytes in return", res->len);
+			if (rsa_block_remain && block_num + 1 == rsa_block_cnt) {
+				g_byte_array_remove_range (res,
+							   rsa_block_remain,
+							   res->len - rsa_block_remain);
+			}
+			for (guint i = 0 ; i < res->len / 2 ; i++) {
+				guint8 tmp = res->data[i];
+				res->data[i] = res->data[res->len - i - 1];
+				res->data[res->len - i - 1] = tmp;
+			}
+			if (rsa_block_remain && block_num + 1 == rsa_block_cnt) {
+				g_byte_array_prepend (rsadump, res->data, rsa_block_remain);
+			} else {
+				g_byte_array_prepend (rsadump, res->data, res->len);
+			}
+		}
+		if (rsa_pubkey_len != rsadump->len) {
+			if (retries++ > 2) {
+				g_set_error (error,
+					     G_IO_ERROR, G_IO_ERROR_FAILED,
+					     "RSA public key length not matched %u: after %u retries: ",
+					     rsadump->len, retries);
+				return FALSE;
+			}
+			g_byte_array_set_size (rsadump, 0);
+			continue;
+		}
+
+		/* success */
+		break;
+	}
+	if (g_getenv ("FWUPD_SYNAPTICS_RMI_VERBOSE") != NULL) {
+		fu_common_dump_full (G_LOG_DOMAIN, "RSA public key",
+				     rsadump->data, rsadump->len,
+				     16, FU_DUMP_FLAGS_NONE);
+	}
+
+	/* sanity check size */
+	if (rsa_pubkey_len != rsadump->len) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "RSA public key length did not match: %u != %u: ",
+			     rsa_pubkey_len, rsadump->len);
+		return FALSE;
+	}
+
+	/* hash firmware data */
+	hash_data = gnutls_malloc (hash_length);
+	gnutls_hash_init (&sha2, GNUTLS_DIG_SHA256);
+	gnutls_hash (sha2, g_bytes_get_data (payload, NULL), g_bytes_get_size (payload));
+	gnutls_hash_deinit (sha2, hash_data);
+
+	/* hash */
+	hash.size = hash_length;
+	hash.data = hash_data;
+
+	/* modulus */
+	m.size = rsadump->len;
+	m.data = rsadump->data;
+
+	/* exponent */
+	e.size = sizeof(exponent);
+	e.data = exponent;
+
+	/* signature */
+	sig.size = g_bytes_get_size (signature);
+	sig.data = (guint8 *) g_bytes_get_data (signature, NULL);
+
+	gnutls_pubkey_init (&pub);
+	ec = gnutls_pubkey_import_rsa_raw (pub, &m, &e);
+	if (ec < 0) {
+		g_prefix_error (error, "failed to import RSA key: ");
+		return FALSE;
+	}
+	ec = gnutls_pubkey_verify_hash2 (pub, GNUTLS_SIGN_RSA_SHA256,
+					 0, &hash, &sig);
+	if (ec < 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "failed to verify firmware");
+		return FALSE;
+	}
+#endif
+	/* IMPORTANT: the signature test is performed on the device *AS WELL*
+	 * and checking the signature here just verifies the firmware will
+	 * flash correctly rather than leave the device in bootloader mode */
+	return TRUE;
```

Comment:
So basically you can write a malware or unsigned payload to the device, but if the signature doesn't match, it will just leave the device in bootloader mode rather than flash it?

Unless we automatically kick it back out of bootloader mode this seems like it enables a denial of service attack to have a fallback code.  
1. If a package is compiled without gnutls support
2. Bad payload is put into a CAB file for local installation (not from LVFS or any remote)
3. Process causes CAB file to be installed
4. Device gets stuck in bootloader mode.

We either need to make sure it gets back out of bootloader mode on it's own, or we should make gnutls a hard requirement for this plugin.

Reply:
> but if the signature doesn't match, it will just leave the device in bootloader mode

@blueue -- do you know this?

Reply:
In my experiment, if the function secure_check returns FALSE, fwupd will call attach to let device be back to normal mode.

Reply:
@blueue I think what mario was asking, was "_if you flash a device that requires an RSA signature test with a file that doesn't have a signature (e.g. if GnuTLS isn't available to test the file before we write it, or if you comment out the code) what happens to the end-user device?_"

Reply:
If the device has RSA public key (you can call it requires RSA signature test) with firmware image file doesn't have a signature, we would treat this update process is invalid. We check this condition in write_firmware before call secure_check function, and return FALSE as failure. That means we won't proceed the firmware update in this condition.

Reply:
But is the device remaining in bootloader mode at the end of that?  Or will the process return back to runtime mode on it's own?

Reply:
The process would call attach function after got FALSE in return from secure_check, so the device would be back to runtime mode in this case.

Reply:
`Well that actually makes my argument stronger.  If the daemon is not linked with gnutls and a unsigned or tampered payload is sent then this function would return `TRUE`.  This means the device gets stuck in bootloader mode since attach is only called in the `FALSE` condition?

If this is the case I think we really should only offer this plugin if the daemon links with gnutls to avoid such a situation.

Reply:
Does the daemon is not lined with gnutls means HAVE_GNUTLS is not defined? It will return TRUE directly if my understanding is correct. If we got TRUE in return of this function, the process would be continued like erasing, flashing, and attaching. In this case, the verification on tool side is not activated, and we will rely on checking scheme on firmware side only (checking signature tool writes to firmware after writes data).

Reply:
Yeah that's the exact scenario I want to make sure that we fully comprehend and it's safe.  If we make gnutls optional for this plugin we need to worry about what happens if someone doesn't have it.

So if payload is unsigned, will some other function call from fwupd later fail because of firmware side checking scheme?  And in that case does the device get attach called again?

Reply:
We can know whether the payload is signed in setup stage because we can get signature flag and size in the header of payload if it is signed. We can know there is an unsigned payload here because it is without signature flag and size in write_firmware function, and return FALSE immediately. And the attach function would be called  after that.

Reply:
Okay but what if it's a payload with a header that looks correct? Or it's signed with an untrusted key?

Reply:
I think it's probably best if I just make the plugin hard depend on gnutls.

Reply:
Done!

Reply:
@superm1 Do you mean payload has signature with untrusted key?  We get public key modulus from device, and import a RSA public key by modulus and exponent. The public key and signature should be paired, so it turns verification failure if the signature is faked.